text/0000-element-internals.md

```diff
@@ -0,0 +1,235 @@
+2. **Form association.** A custom element can communicate its states (e.g. valid/invalid, checked/unchecked) as part of a [form](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form), as well as use lifecycle callbacks corresponding to the form itself (form associated, form reset, etc.).
+
+## Detailed design
+
+In web component frameworks such as [Lit](https://lit.dev/), [Stencil](https://stenciljs.com/), and [FAST](https://www.fast.design/), component authors have access to the `class` that is registered with `customElements.define()`, as well as the `HTMLElement` instance via `this`. In LWC, the framework mediates these interactions, so we must provide an explicit API for these features.
+
+Additionally, calling `attachInternals()` is inherently risky, because it can unintentionally leak private data. For instance, [it is possible](https://codepen.io/nolanlawson-the-selector/pen/gOdWPQJ) with vanilla custom elements to call `attachInternals()` from outside a component and gain access to its (closed) shadow root:
+
+```js
+const { shadowRoot } = document.querySelector('my-component').attachInternals();
+```
+
+This pitfall should be avoided in LWC's implementation.
+
+### Accessing ElementInternals
+
+Within an LWC component, component authors can call `this.attachInternals()` to create an `ElementInternals` object:
+
+```js
+export default class extends LightningElement {
+  constructor() {
+    super();
+    const internals = this.attachInternals();
+    internals.role = 'button';
+  }
+}
+```
+
+Most commonly, developers will probably assign the internals to a class property:
+
+```js
+export default class extends LightningElement {
+  internals;
+  constructor() {
+    super();
+    this.internals = this.attachInternals();
+    this.internals.role = 'button';
+  }
+}
+```
+
+`this.attachInternals` is available at any point in the custom element lifecycle, similar to `this.template`. It can also be directly called from a class property:
+
+```js
+export default class extends LightningElement {
+  internals = this.attachInternals();
+}
+```
+
+(Note that vanilla custom elements allow this as well.)
+
+To avoid leaking private component details, accessing internals from outside a component is not allowed:
+
+```js
+const component = document.querySelector('my-component');
+console.log(component.attachInternals); // undefined
+```
+
+`this.attachInternals()` can only be called once, and will throw an error if called twice. (This is how vanilla custom elements work.) So component authors are encouraged to guard against this by avoiding calling `this.attachInternals()` in a lifecycle hook that may be called multiple times, such as `connectedCallback` or `renderedCallback`.
+
+In Server-Side Rendering (SSR), `this.attachInternals` will be undefined.

```

Comment:
Maybe this should throw an explicit error instead in SSR? This might be a better dev experience.